tests/sentry/ratelimits/test_cardinality.py

```diff
@@ -0,0 +1,163 @@
+from typing import Collection, Optional, Sequence
+
+import pytest
+
+from sentry.ratelimits.cardinality import Quota, RedisCardinalityLimiter, RequestedQuota
+
+
+@pytest.fixture
+def limiter():
+    return RedisCardinalityLimiter()
+
+
+class LimiterHelper:
+    """
+    Wrapper interface around the rate limiter, with specialized, stateful and
+    primitive interface for more readable tests.
+    """
+
+    def __init__(self, limiter: RedisCardinalityLimiter):
+        self.limiter = limiter
+        self.quotas = [Quota(window_seconds=3600, granularity_seconds=60, limit=10)]
+        self.timestamp = 3600
+
+    def add_value(self, value: int) -> Optional[int]:
+        values = self.add_values([value])
+        if values:
+            (value,) = values
+            return value
+
+    def add_values(self, values: Sequence[int]) -> Collection[int]:
+        request = RequestedQuota(prefix="hello", unit_hashes=values, quotas=self.quotas)
+        new_timestamp, grants = self.limiter.check_within_quotas(
+            [request], timestamp=self.timestamp
+        )
+        self.limiter.use_quotas(grants, new_timestamp)
+        (grant,) = grants
+        return grant.granted_unit_hashes
+
+
+def test_basic(limiter: RedisCardinalityLimiter):
+    helper = LimiterHelper(limiter)
+
+    for _ in range(20):
+        assert helper.add_value(1) == 1
+
+    for _ in range(20):
+        assert helper.add_value(2) == 2
+
+    assert [helper.add_value(10 + i) for i in range(100)] == list(range(10, 18)) + [None] * 92
+
+    helper.timestamp += 3600
+
+    # some old keys 1, 2 expired, we should be able to admit the same keys
+    # 10..18 again + 2 new keys
```

Comment:
I don't think this is what is happening. 
By bumping the time ahead of a full window, all the keys above are expired, this is why you allow all the keys from 10 to 20

Reply:
fixed